File: Hashing 1/subarraySumZero.py
def checkSubarrayWithZeroSum(arr):
    s = set()

    prefixSum = arr[0]
    s.add(prefixSum)

    for i in range(1, len(arr)):
        prefixSum += arr[i]
        s.add(prefixSum)

    if len(s) < len(arr) or (0 in s):
        return True

    return False


print(checkSubarrayWithZeroSum([2, 5, 6, -111, 10, -4]))

File: Hashing 1/SubarraySumZero.java
import java.util.HashSet;

public class SubarraySumZero {
    static boolean checkSubarrayWithZeroSum(int[] arr) {
        HashSet<Integer> s = new HashSet<Integer>();

        int prefixSum = arr[0];
        s.add(prefixSum);

        for (int i = 1; i < arr.length; i++) {
            prefixSum += arr[i];
            s.add(prefixSum);
        }

        if (s.size() < arr.length || s.contains(0)) {
            return true;
        }

        return false;
    }

    public static void main(String[] args) {
        int[] arr = { 2, 5, 6, -2, 0, -4 };

        System.out.println(checkSubarrayWithZeroSum(arr));
    }
}
